### PR TITLE
Refactoring to avoid cyclic dependency

### PR DESCRIPTION
**Pull request by kunalpal**
Hi. Earlier, the well-used functions `MODULE_LIST` and `SCRIPT_LIST` were in `__init__.py` which caused cyclic dependencies throughout other imports (e.g. `lib/compile.py -> lib/templates.py -> __init__.py -> lib/compile.py`). Since, these functions are used everywhere it seemed reasonable to refactor them (into `lib/scripts.py`) to avoid these problems. As suggested in PR #814, I am creating this separate PR for better review and seamless merging.

### CODE REVIEW

**Comment by henrykironde** [comment]
@kunalpal, amend the commit message explaining the major changes, examples [ 1](https://github.com/weecology/retriever/commit/a9208eaec63107e5216543523e4bcded547f231d) [2](https://github.com/weecology/retriever/commit/4702ed05c081af938baa12309f2e8bb3f52964ad) [3](https://github.com/weecology/retriever/commit/7cfa9a72d269630cb3952ed2cfb6d475ca7fe373)


Another cyclic dependency to look at would be in tools https://github.com/weecology/retriever/blob/master/lib/tools.py#L107. I will review this and give you feed back. Basing on the number of PR's we may take more time since the changes are in some sensitive parts.

**Comment by kunalpal** [comment]
Sure. I am looking into [tools.py](https://github.com/weecology/retriever/blob/master/lib/tools.py#L107). I am also changing the commit messages to be more verbose.

File: docs/conf.py [hunk 1]
```diff
@@ -9,9 +9,9 @@
     sys.setdefaultencoding('latin-1')
 
 from builtins import str
-from retriever import VERSION,COPYRIGHT
+from retriever.lib.defaults import VERSION,COPYRIGHT
```
Comment (ethanwhite) [inline docs/conf.py]: Missing space after `,`

File: __init__.py [hunk 2]
```diff
@@ -119,57 +92,6 @@ def to_str(object, object_encoding=sys.stdout):
...
-    return modules
-
-
-def SCRIPT_LIST(force_compile=False):
-    return [module.SCRIPT for module in MODULE_LIST(force_compile)]
-
-
 def ENGINE_LIST():
     from retriever.engines import engine_list
     return engine_list
```
Comment (ethanwhite) [inline __init__.py]: What is your thinking on not moving `ENGINE_LIST()` to defaults?

Comment (kunalpal) [inline __init__.py]: Hi @ethanwhite. I have fixed the typo in `conf.py`. Additionally, this PR was a part of #814 and I only suffered from the cyclic dependency in `SCRIPT_LIST` and `MODULE_LIST`. However, the function `ENGINE_LIST` seems a placeholder for fetching `engine_list` from `retriever.engine`. I have refactored ENGINE_LIST now as well. Could you please check and give me your feedback..:)

**Comment by kunalpal** [comment]
Hi @ethanwhite. I have fixed the typo in conf.py. Additionally, this PR was a part of #814 and I only suffered from the cyclic dependency in SCRIPT_LIST and MODULE_LIST. However, the function ENGINE_LIST seems a placeholder for fetching engine_list from retriever.engine. I have refactored ENGINE_LIST now as well. Could you please check and give me your feedback..:)

**Comment by kunalpal** [comment]
Hi @henrykironde. I have also resolved the cyclic dependency in `retriever.lib.tools` by moving the `choose_engine` function into `retriever.engines.___init__.py`. Looking forward to your remarks..:)

**Comment by henrykironde** [comment]
Thanks, I will need more time to carefully go through this PR Since we are changing core components. That applies to your other PR that removes python scripts

**Comment by kunalpal** [comment]
Thanks @henrykironde. No problem.

File: retriever/__main__.py [hunk 3]
```diff
@@ -22,10 +22,12 @@
 if hasattr(sys, 'setdefaultencoding'):
     sys.setdefaultencoding(encoding)
 
-from retriever import VERSION, SCRIPT_LIST, HOME_DIR, sample_script, CITATION
-from retriever.engines import engine_list
+from retriever import sample_script, CITATION
+from retriever.lib.defaults import VERSION, HOME_DIR
```
Comment (henrykironde) [inline retriever/__main__.py]: VERSION in` from retriever.lib.defaults import VERSION, HOME_DIR` not being used

File: scripts/MammalSuperTree.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #retriever
-from retriever import VERSION
+from retriever.lib.defaults import VERSION
```
Comment (henrykironde) [inline scripts/MammalSuperTree.py]: Remove unused imports

File: scripts/breed_bird_survey.py [hunk 5]
```diff
@@ -14,8 +14,8 @@
 from decimal import Decimal
 from retriever.lib.templates import Script
 from retriever.lib.models import Table, Cleanup, no_cleanup, correct_invalid_value
-from retriever import HOME_DIR, open_fr, open_fw
-
+from retriever import open_fr, open_fw
+from retriever.lib.defaults import HOME_DIR
```
Comment (henrykironde) [inline scripts/breed_bird_survey.py]: Remove unused imports

File: scripts/breed_bird_survey_50stop.py [hunk 6]
```diff
@@ -15,7 +15,8 @@
 from decimal import Decimal
 from retriever.lib.templates import Script
 from retriever.lib.models import Table, Cleanup, no_cleanup, correct_invalid_value
-from retriever import HOME_DIR, open_fr, open_fw
+from retriever import open_fr, open_fw
+from retriever.lib.defaults import HOME_DIR
```
Comment (henrykironde) [inline scripts/breed_bird_survey_50stop.py]: Remove unused imports

File: scripts/plant_life_hist_eu.py [hunk 7]
```diff
@@ -3,7 +3,8 @@
 from builtins import str
 from retriever.lib.models import Table, Cleanup, correct_invalid_value
 from retriever.lib.templates import Script
-from retriever import HOME_DIR, open_fr, open_fw
+from retriever import open_fr, open_fw
+from retriever.lib.defaults import HOME_DIR
```
Comment (henrykironde) [inline scripts/plant_life_hist_eu.py]: Remove unused imports

File: scripts/wood_density.py [hunk 8]
```diff
@@ -12,7 +12,8 @@
 from retriever.lib.templates import Script
 from retriever.lib.models import Table
 from retriever.lib.excel import Excel
-from retriever import HOME_DIR, open_fr, open_fw, open_csvw, to_str
+from retriever import open_fr, open_fw, open_csvw, to_str
+from retriever.lib.defaults import HOME_DIR
```
Comment (henrykironde) [inline scripts/wood_density.py]: Remove unused imports

File: retriever/__main__.py [hunk 9]
```diff
@@ -22,10 +22,12 @@
...
+from retriever.lib.scripts import SCRIPT_LIST
+from retriever.engines import engine_list, choose_engine
 from retriever.lib.repository import check_for_updates
-from retriever.lib.tools import choose_engine, name_matches, reset_retriever
+from retriever.lib.tools import name_matches, reset_retriever
```
Comment (henrykironde) [inline retriever/__main__.py]: lets add one line before main ``def main():``

File: retriever/__main__.py [hunk 10]
```diff
@@ -22,10 +22,12 @@
 if hasattr(sys, 'setdefaultencoding'):
     sys.setdefaultencoding(encoding)
 
-from retriever import VERSION, SCRIPT_LIST, HOME_DIR, sample_script, CITATION
-from retriever.engines import engine_list
+from retriever import sample_script, CITATION
+from retriever.lib.defaults import VERSION, HOME_DIR
+from retriever.lib.scripts import SCRIPT_LIST
+from retriever.engines import engine_list, choose_engine
```
Comment (henrykironde) [inline retriever/__main__.py]: Lets add one line before main ``def main():``

**Comment by henrykironde** [comment]
@kunalpal, I have made some review comments, Look through them  and here more explanations

 
Remove import `from retriever.lib.defaults import VERSION`

Update your branch with master, the script npn.py has DATA_WRITE_PATH,
Since you have changed this to defaults, The script will fail. Do you have anyway to avoid this kind of problem?

If we change the constants or defaults from `retriever/__init__.py` to `retriever/lib/defaults.py`, then all scripts having these values will fail.

This means that if am using 2.0.dev and I get the latest scripts after this commit is merged, they will fail because they will be looking for defaults in the old path ./__init__.py`   instead of `retriever/__init__.py`

I have looked through the scripts and we have
https://github.com/weecology/retriever/pull/825/files#diff-96563a14c0b6113df4da02deac07cecdL6

```
from retriever import HOME_DIR, open_fr, open_fw (old)
from retriever.lib.defaults import HOME_DIR(new) 
```
So my old version of 2.o.dev fails due to this.

Some scripts have the import statements but do not use them, 
So removing the imports in the new scripts is fine


In main we have VERSION not used, Lets remove it
`from retriever.lib.defaults import VERSION, ....`

Also make sure you test `retriever ls` 

This could give you an idea what could be happening to the your PR *_Avoid creation of intermediate python scripts_* and how we could solve it's issues

File: retriever/lib/defaults.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+SCRIPT_WRITE_PATH = SCRIPT_SEARCH_PATHS[-1]
+DATA_SEARCH_PATHS = [
+    "./",
+    "{dataset}",
+    "raw_data/{dataset}",
+    os.path.join(HOME_DIR, 'raw_data/{dataset}'),
+]
+DATA_WRITE_PATH = DATA_SEARCH_PATHS[-1]
+
+# Create default data directory
```
Comment (henrykironde) [inline retriever/lib/defaults.py]: Any Reason we are not including 
CITATION
and 
sample_script 
in defaults?

**Comment by henrykironde** [comment]
Closing because of many conflicts due to new changes and new added procedures. We will refer to this PR incase we decide to proceed in this path.
@kunalpal we really thank you for the great work.
@ShivamNegi refer to this work. This is an old branch that kunalpal was working on, but outdated now.

**Pull request closed without merging**